File: python/ppml/example/benchmark/fgboost_benchmark/fgboost_real_data_benchmark.py
#
# Copyright 2016 The BigDL Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import numpy as np
import pandas as pd
import argparse
import time
from bigdl.ppml.fl import *
from bigdl.ppml.fl.data_utils import *
from bigdl.ppml.fl.algorithms.fgboost_regression import FGBoostRegression
from urllib.parse import urlparse
from os.path import exists
from bigdl.dllib.utils import log4Error
import os
import sys


def is_local_and_existing_uri(uri):
    parsed_uri = urlparse(uri)

    log4Error.invalidInputError(not parsed_uri.scheme or parsed_uri.scheme == 'file',
                                "Not Local File!")

    log4Error.invalidInputError(not parsed_uri.netloc or parsed_uri.netloc.lower() == 'localhost',
                                "Not Local File!")

    log4Error.invalidInputError(exists(parsed_uri.path),
                                "File Not Exist!")


if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        prog='PPML FGBoost Benchmark',
        description='Run PPML FGBoost Benchmark using real data')
    parser.add_argument('--train_path',
                        type=str,
                        help='The train data path.')
    parser.add_argument('--test_path',
                        type=str,
                        help='The test data path.')
    parser.add_argument('--data_size',
                        type=int,
                        default=1,
                        help='The size of data copy, e.g. 10 means using 10 copies of data.')
    parser.add_argument('--num_round',
                        type=int,
                        default=10,
                        help='The boosting rounds.')
    args = parser.parse_args()
    # process path traversal issue
    safe_dir_train = "/safe_dir/"
    dir_name_train = os.path.dirname(args.train_path)
    if '../' in dir_name_train:
        sys.exit(1)
    safe_dir_train = dir_name_train
    file_name_train = os.path.basename(args.train_path)
    train_path = os.path.join(safe_dir_train, file_name_train)
    
    safe_dir_test = "/safe_dir/"
    dir_name_test = os.path.dirname(args.test_path)
    if '../' in dir_name_test:
        sys.exit(1)
    safe_dir_test = dir_name_test
    file_name_test = os.path.basename(args.test_path)
    test_path = os.path.join(safe_dir_test, file_name_test)

    init_fl_context(1)
    is_local_and_existing_uri(train_path)
    df_train = pd.read_csv(train_path)
    fgboost_regression = FGBoostRegression()
    
    df_x = df_train.drop('SalePrice', 1)
    df_y = df_train.filter(items=['SalePrice'])
    x = convert_to_numpy(df_x)
    y = convert_to_numpy(df_y)
    x_stacked = []
    y_stacked = []
    for i in range(args.data_size):
        x_stacked.append(x)
        y_stacked.append(y)
    
    x_stacked = np.array(x_stacked)
    y_stacked = np.array(y_stacked)
    
    fgboost_regression.fit(x_stacked.reshape(-1, x_stacked.shape[-1]),
                           y_stacked.reshape(-1, y_stacked.shape[-1]),
                           num_round=args.num_round)
    
    is_local_and_existing_uri(test_path)
    df_test = pd.read_csv(test_path)
    result = fgboost_regression.predict(df_test, feature_columns=df_x.columns)
    

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/fl/example/benchmark/FGBoostRealDataBenchmark.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.bigdl.ppml.fl.example.benchmark

import com.intel.analytics.bigdl.dllib.tensor.Tensor
import com.intel.analytics.bigdl.ppml.fl.algorithms.FGBoostRegression
import com.intel.analytics.bigdl.ppml.fl.FLContext
import com.intel.analytics.bigdl.ppml.fl.fgboost.common.XGBoostFormatValidator
import com.intel.analytics.bigdl.ppml.fl.utils.{DataFrameUtils, TimingSupportive}
import scopt.OptionParser

import scala.io.Source
import scala.util.Random


object FGBoostRealDataBenchmark extends TimingSupportive {
  case class Params(trainPath: String = null,
                    testPath: String = null,
                    dataSize: Int = 1,
                    numRound: Int = 10)
  def main(args: Array[String]): Unit = {
    val parser = new OptionParser[Params]("Text Classification Example") {
      opt[String]("trainPath")
        .text("data size")
        .action((x, params) => params.copy(trainPath = x))
        .required()
      opt[String]("testPath")
        .text("data size")
        .action((x, params) => params.copy(testPath = x))
        .required()
      opt[Int]("dataSize")
        .text("data dimension")
        .action((x, params) => params.copy(dataSize = x))
      opt[Int]("numRound")
        .text("boosting round numer")
        .action((x, params) => params.copy(numRound = x))
    }

    val param = parser.parse(args, Params()).get
    val spark = FLContext.getSparkSession()
    val trainDf = spark.read.option("header", "true").csv(param.trainPath)
    val testDf = spark.read.option("header", "true").csv(param.testPath)


    // Data copy for simulating large dataset
    val trainDataArray = DataFrameUtils.toTensorArray(trainDf.drop("SalePrice"))
    val trainLabelArray = DataFrameUtils.toArray(trainDf, "SalePrice")

    var trainDataStacked = Array[Tensor[Float]]()
    var trainLabelStacked = Array[Float]()
    (0 until param.dataSize).foreach(_ => {
      trainDataStacked = trainDataStacked ++ trainDataArray
      trainLabelStacked = trainLabelStacked ++ trainLabelArray
    })

    FLContext.initFLContext(1)
    val fGBoostRegression = new FGBoostRegression(
      learningRate = 0.1f, maxDepth = 7, minChildSize = 5)
    fGBoostRegression.fit(
      trainDataStacked,
      trainLabelStacked,
      param.numRound)

    val fGBoostResult = fGBoostRegression.predict(DataFrameUtils.toTensorArray(testDf))
      .map(tensor => tensor.value())
      .map(math.exp(_))

  }
}
